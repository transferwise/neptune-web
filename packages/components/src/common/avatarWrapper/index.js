import AvatarWrapper from './AvatarWrapper';

export default AvatarWrapper;
