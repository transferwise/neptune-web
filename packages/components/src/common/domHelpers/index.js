export function isTouchDevice() {
  return !!window?.matchMedia('(pointer: coarse)')?.matches;
}

export {
  addClickClassToDocumentOnIos,
  removeClickClassFromDocumentOnIos,
} from './documentIosClick';
