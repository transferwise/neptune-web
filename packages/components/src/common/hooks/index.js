export { default as useHasIntersected } from './useHasIntersected';
export { default as useClientWidth } from './useClientWidth';
export { default as useRTL } from './useRTL';
