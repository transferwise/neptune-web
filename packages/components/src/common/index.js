export { default as HistoryNavigator } from './historyNavigator';
export { addNoScrollBodyClass, removeNoScrollBodyClass } from './DOMOperations';
export { Theme } from './theme';
// Props
export { Breakpoint } from './propsValues/breakpoint';
export { Size } from './propsValues/size';
export { Type } from './propsValues/type';
export { DateMode } from './propsValues/dateMode';
export { MonthFormat } from './propsValues/monthFormat';
// Position and Placement are duplicated and needs to be aligned with a breaking change
export { Position } from './propsValues/position';
export { Layout } from './propsValues/layouts';
export { FormControlType } from './propsValues/formControlType';
export { Status } from './propsValues/status';
export { Sentiment } from './propsValues/sentiment';
export { ProfileType } from './propsValues/profileType';
export { Key } from './key';
