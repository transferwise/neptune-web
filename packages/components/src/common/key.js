export const Key = {
  ENTER: 'Enter',
  SPACE: [' ', 'Spacebar'],
};
