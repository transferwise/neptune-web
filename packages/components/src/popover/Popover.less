@import (reference) '@transferwise/neptune-css/src/less/variables/_typography.less';
@import (reference) '@transferwise/neptune-css/src/less/core/_typography.less';
// This can't be wrapped in np-popover because styles won't get applied to bottomsheet
//which is opened in a portal hence it won't be contained in np-popover.
.np-popover__container {
  .np-popover__content {
    font-size: @font-size-small;
  }

  &.np-panel {
    .np-popover__title {
      .h5;
    }

    .np-popover__content {
      padding: var(--size-16) var(--size-24);
      max-width: 276px;
    }
  }

  &.np-bottom-sheet {
    .np-popover__title {
      .h2;
    }

    .np-popover__content {
      padding: var(--size-32) var(--size-24);
    }
  }
}
