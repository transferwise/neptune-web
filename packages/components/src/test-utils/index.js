import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
