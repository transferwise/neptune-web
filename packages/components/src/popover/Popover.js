import React, { useRef } from 'react';
import Types from 'prop-types';
import classnames from 'classnames';

import { logActionRequiredIf, deprecated } from '../utilities';
import { Position } from '../common';
import ResponsivePanel from '../common/responsivePanel';
import { useAttachEvent } from '../common/hooks';
import keyCodes from '../common/keyCodes';

import './Popover.css';

const expiryDate = new Date('17-08-2021');

const Popover = ({
  children,
  classNames,
  content,
  onClose,
  open,
  preferredPlacement,
  position = preferredPlacement,
  title,
}) => {
  logActionRequired({ position });

  const triggerRef = useRef(null);
  const responsivePanelRef = useRef(null);

  const callBack = (val) => val && onClose(true);

  useAttachEvent({
    eventType: 'click',
    condition: (event) =>
      ![responsivePanelRef, triggerRef].some((el) => el?.current?.contains(event.target)),
    callBack,
    attachListener: open,
  });

  useAttachEvent({
    eventType: 'keydown',
    condition: (event) => event.keyCode === keyCodes.ESCAPE,
    callBack,
    attachListener: open,
  });

  return (
    <span className={classnames('np-popover', classNames)}>
      <span className="d-inline-block" ref={triggerRef}>
        {children}
        {open && (
          <span role="status" className="sr-only">
            {title}
            {content}
          </span>
        )}
      </span>
      <ResponsivePanel
        open={open}
        ref={responsivePanelRef}
        triggerRef={triggerRef}
        position={deprecatedPositions[position] || position}
        arrow
      >
        <div className="np-popover__content" aria-hidden={!open} role="tooltip">
          <div className="h5">{title}</div>
          {content}
        </div>
      </ResponsivePanel>
    </span>
  );
};

const logActionRequired = ({ position }) => {
  logActionRequiredIf(
    `Popover has deprecated the ${position} value for the 'position' prop. Please use ${deprecatedPositions[position]} instead.`,
    deprecatedPositions[position],
  );
};

Popover.Position = {
  BOTTOM: Position.BOTTOM,
  BOTTOM_LEFT: Position.BOTTOM_LEFT,
  BOTTOM_RIGHT: Position.BOTTOM_RIGHT,
  LEFT: Position.LEFT,
  LEFT_TOP: Position.LEFT_TOP,
  RIGHT: Position.RIGHT,
  RIGHT_TOP: Position.RIGHT_TOP,
  TOP: Position.TOP,
};

Popover.defaultProps = {
  classNames: undefined,
  onClose: undefined,
  open: false,
  position: Popover.Position.TOP,
  title: undefined,
};

Popover.propTypes = {
  children: Types.node.isRequired,
  classNames: Types.string,
  content: Types.node.isRequired,
  onClose: Types.func,
  open: Types.bool,
  /** @DEPRECATED please use position instead */
  // eslint-disable-next-line
  preferredPlacement: deprecated(
    Types.oneOf([
      Popover.Position.TOP,
      Popover.Position.RIGHT,
      Popover.Position.BOTTOM,
      Popover.Position.LEFT,
      Popover.Position.LEFT_TOP,
      Popover.Position.RIGHT_TOP,
      Popover.Position.BOTTOM_RIGHT,
      Popover.Position.BOTTOM_LEFT,
    ]),
    {
      component: 'Popover',
      newProp: 'position',
      expiryDate,
    },
  ),
  position: Types.oneOf([
    Popover.Position.BOTTOM,
    Popover.Position.LEFT,
    Popover.Position.RIGHT,
    Popover.Position.TOP,
    /* @DEPRECATED Please use BOTTOM instead. */
    Popover.Position.BOTTOM_LEFT,
    /* @DEPRECATED Please use BOTTOM instead. */
    Popover.Position.BOTTOM_RIGHT,
    /* @DEPRECATED Please use TOP instead. */
    Popover.Position.LEFT_TOP,
    /* @DEPRECATED Please use TOP instead. */
    Popover.Position.RIGHT_TOP,
  ]),
  title: Types.node,
};

export const deprecatedPositions = {
  [Position.BOTTOM_LEFT]: Popover.Position.BOTTOM,
  [Position.BOTTOM_RIGHT]: Popover.Position.BOTTOM,
  [Position.LEFT_TOP]: Popover.Position.TOP,
  [Position.RIGHT_TOP]: Popover.Position.TOP,
};

export default Popover;
