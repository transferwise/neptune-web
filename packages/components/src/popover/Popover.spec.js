import React from 'react';
import { fireEvent, render, waitFor, screen } from '../test-utils';
import Popover from './Popover';
import { Breakpoint } from '../common';

jest.mock('../common/bottomSheet', () => {
  // eslint-disable-next-line
  const { forwardRef } = require('react');
  return forwardRef((props, ref) => {
    // eslint-disable-next-line react/prop-types
    const { open, children } = props;
    return open ? (
      <div ref={ref} className="np-bottom-sheet">
        {children}
      </div>
    ) : null;
  });
});

describe('Popover', () => {
  const props = {
    arrow: true,
    content: 'content',
    open: false,
    position: Popover.Position.BOTTOM,
    title: 'title',
    clickOutsideCallback: jest.fn(),
  };

  let container;
  let rerender;

  describe('on desktop', () => {
    it('renders correctly when open is true', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props} open>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(container).toMatchSnapshot();
    });

    it('renders correctly when open is false', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props}>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(container).toMatchSnapshot();
    });

    it('renders Panel based on open', async () => {
      await waitFor(() => {
        ({ container, rerender } = render(
          <Popover {...props}>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getPanel()).not.toBeInTheDocument();

      await waitFor(() => {
        ({ container, rerender } = render(
          <Popover {...props} open>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getPanel()).toBeInTheDocument();
    });
  });

  describe('on mobile', () => {
    beforeEach(() => {
      global.innerWidth = Breakpoint.SMALL - 1;
    });

    it('renders', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props} initialOpen>
            <button type="button">Open</button>
          </Popover>,
        ));
      });
      expect(container).toMatchSnapshot();
    });

    it('renders BottomSheet based on open', async () => {
      await waitFor(() => {
        ({ container, rerender } = render(
          <Popover {...props} open>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getBottomSheet()).toBeInTheDocument();

      await waitFor(() => {
        rerender(
          <Popover {...props} open={false}>
            <button type="button">Open</button>
          </Popover>,
        );
      });

      expect(getBottomSheet()).not.toBeInTheDocument();
    });

    it('calls clickOutsideCallback withg correct value', async () => {
      await waitFor(() => {
        ({ container, rerender } = render(
          <Popover {...props} open>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      fireEvent.click(screen.getByText('Open'));
      expect(props.clickOutsideCallback).not.toHaveBeenCalled();

      fireEvent.click(document);
      expect(props.clickOutsideCallback).toHaveBeenCalledWith(true);
    });
  });

  const getPanel = () => document.querySelector('.np-panel--open');
  const getBottomSheet = () => document.querySelector('.np-bottom-sheet');
});
