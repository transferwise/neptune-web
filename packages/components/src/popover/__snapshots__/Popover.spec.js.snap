// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover on desktop renders correctly when open is true 1`] = `
<div>
  <span
    class="np-popover"
  >
    <span
      class="d-inline-block"
    >
      <button
        type="button"
      >
        Open
      </button>
      <span
        class="sr-only"
        role="status"
      >
        title
        content
      </span>
    </span>
    <div
      class="np-size-swapper d-flex"
      style="visibility: visible;"
    >
      <div
        class="np-panel np-panel--open"
        data-popper-escaped="true"
        data-popper-placement="bottom"
        data-popper-reference-hidden="true"
        style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 16px);"
      >
        <div
          class="np-panel__content"
        >
          <div
            aria-hidden="false"
            class="np-popover__content"
            role="tooltip"
          >
            <div
              class="h5"
            >
              title
            </div>
            content
          </div>
          <div
            class="np-panel__arrow"
            style="position: absolute; left: 0px; transform: translate(5px, 0px);"
          />
        </div>
      </div>
    </div>
  </span>
</div>
`;

exports[`Popover on mobile renders 1`] = `
<div>
  <span
    class="np-popover"
  >
    <span
      class="d-inline-block"
    >
      <button
        type="button"
      >
        Open
      </button>
      <span
        class="sr-only"
        role="status"
      >
        title
        content
      </span>
    </span>
    <div
      class="np-size-swapper d-flex"
      style="visibility: visible;"
    >
      <div
        class="np-bottom-sheet"
      >
        <div
          aria-hidden="false"
          class="np-popover__content"
          role="tooltip"
        >
          <div
            class="h5"
          >
            title
          </div>
          content
        </div>
      </div>
    </div>
  </span>
</div>
`;
